test/e2e/network/endpointslice.go

```diff
@@ -299,6 +300,113 @@ var _ = SIGDescribe("EndpointSlice", func() {
+		createTestPods(podClient, 0, 100, sharedLabels)
+
+		// Ensure EndpointSlice has been created for Service and has at least 50
+		// endpoints. Not waiting for all Pods to be part of EndpointSlice to
+		// ensure EndpointSlice controller is even busier in following step.
+		lSelector := discoveryv1beta1.LabelServiceName + "=" + svc.Name
+		err := wait.Poll(5*time.Second, 3*time.Minute, func() (bool, error) {
+			esList, err := cs.DiscoveryV1beta1().EndpointSlices(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
+			framework.ExpectNoError(err, "Error listing EndpointSlices for %s/%s Service", f.Namespace.Name, svc.Name)
+			if len(esList.Items) != 1 {
+				framework.Logf("Expected 1 EndpointSlice for %s/%s Service, got %d", f.Namespace.Name, svc.Name, len(esList.Items))
+				return false, nil
+			}
+
+			if len(esList.Items[0].Endpoints) < 50 {
+				framework.Logf("Expected at least 50 EndpointSlice endpoints for %s/%s Service, got %d", f.Namespace.Name, svc.Name, len(esList.Items[0].Endpoints))
+				return false, nil
+			}
+
+			return true, nil
+		})
+		framework.ExpectNoError(err, "Timed out waiting for EndpointSlice to have at least 50 endpoints for %s/%s Service", f.Namespace.Name, svc.Name)
+
+		// In this step we run 4 goroutines simultaneously. The first 3 are all
+		// trying to keep the EndpointSlice controller as busy as possible.
+		// There are 2 goroutines creating 51 new Pods each at the same time
+		// another goroutine is deleting the 100 Pods created initially. The
+		// final goroutine deletes the Service in the midst of the chaos to
+		// ensure that all of this action in the EndpointSlice controller will
+		// not interfere with garbage collection.
+		var wg sync.WaitGroup
+		wg.Add(4)
+
+		go func() {
+			defer wg.Done()
+			createTestPods(podClient, 100, 151, sharedLabels)
+		}()
+
+		go func() {
+			defer wg.Done()
+			createTestPods(podClient, 200, 251, sharedLabels)
+		}()
+
+		go func() {
+			defer wg.Done()
+			deleteTestPods(podClient, 0, 100, sharedLabels)
+		}()
+
+		go func() {
+			defer wg.Done()
+			// Wait for at least 2 EndpointSlices to exist for this Service.
+			err := wait.Poll(5*time.Second, 3*time.Minute, func() (bool, error) {
+				esList, err := cs.DiscoveryV1beta1().EndpointSlices(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
+				framework.ExpectNoError(err, "Error listing EndpointSlices for %s/%s Service", f.Namespace.Name, svc.Name)
+				if len(esList.Items) < 2 {
+					framework.Logf("Waiting for second EndpointSlice to be created for %s/%s Service", f.Namespace.Name, svc.Name)
+					return false, nil
+				}
+
+				return true, nil
+			})
+			framework.ExpectNoError(err, "Timed out waiting for second EndpointSlice to be created for %s/%s Service", f.Namespace.Name, svc.Name)
+
+			err = cs.CoreV1().Services(f.Namespace.Name).Delete(context.TODO(), svc.Name, metav1.DeleteOptions{})
```

Comment:
We use `metav1.DeletePropagationForeground` as `PropagationPolicy`. If that isn't the default(*) , the service is going to be deleted although there are still some endpointslices owned by it, right?

(*) If I remember (and understood) correctly, that depends on how the ownership reference is set up in the endpointslices. I don't have a cluster running right now, so I can't check that.

Reply:
endpointslices have `blockOwnerDeletion: true`. So does this block deletion of the service, regardless of the `PropagationPolicy`? https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion describes `blockOwnerDeletion` but doesn't say anything about how it is handled when foreground cascading deletion is not used.

That then begs the question whether kubectl used foreground cascading deletion in the test script that I was trying to write. I am not sure anymore because in an experiment with `kubectl delete deployment/service-stress-test`, the deployment was deleted although there was a service which had it as owner with `blockOwnerDeletion: true` and the service itself couldn't be deleted because I had set a finalizer.

I was expecting `kubectl` to use foreground cascading based on https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#setting-the-cascading-deletion-policy, specifically "kubectl also supports cascading deletion. ... The default value for --cascade is true."

But I misunderstood `--cascading`: `kubectl -v=10` shows that `--cascade` only switches between `propagationPolicy":"Orphan"` and `"propagationPolicy":"Background"`. I don't see an option to select `propagationPolicy":"Foreground"`.

After switching to `curl` with `"propagationPolicy":"Foreground"`, I managed to trigger the race condition in my test script. I've posted it in the issue: https://github.com/kubernetes/kubernetes/issues/91287#issuecomment-633494000

Reply:
Thanks for the detailed steps to reproduce this issue! I've added an e2e test that uses a similar approach to this. Although that script will still appear to fail for me on an e2e cluster with this fix, it is because the deployment sometimes sticks around for an extra second or two. In my testing, EndpointSlices are consistently empty in the script output with these fixes.

There is unfortunately no way to completely guarantee that the EndpointSlice controller will not try to create EndpointSlices after Service deletion begins, but this should prevent it from happening as soon as the EndpointSlice controller is aware the Service is being deleted.